exporter/loadbalancingexporter/exporter.go

```diff
@@ -16,42 +16,208 @@ package loadbalancingexporter
+
+		var err error
+		res, err = newDNSResolver(dnsLogger, oCfg.Resolver.DNS.Hostname)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if res == nil {
+		return nil, errNoResolver
+	}
+
 	return &exporterImp{
 		logger: params.Logger,
 		config: *oCfg,
+
+		res: res,
+
+		exporters:            map[string]component.TracesExporter{},
+		exporterFactory:      otlpexporter.NewFactory(),
+		templateCreateParams: tmplParams,
 	}, nil
 }
 
 func (e *exporterImp) Start(ctx context.Context, host component.Host) error {
+	e.res.onChange(e.onBackendChanges)
+	if err := e.res.start(ctx); err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func (e *exporterImp) onBackendChanges(resolved []string) {
+	resolved = sort.StringSlice(resolved)
+	newRing := newHashRing(resolved)
+
+	if !newRing.equal(e.ring) {
+		e.updateLock.Lock()
+		defer e.updateLock.Unlock()
+
+		e.ring = newRing
+
+		// TODO: set a timeout?
+		ctx := context.Background()
+
+		// add the missing exporters first
+		e.addMissingExporters(ctx, resolved)
+		e.removeExtraExporters(ctx, resolved)
+	}
+}
+
+func (e *exporterImp) addMissingExporters(ctx context.Context, endpoints []string) {
+	for _, endpoint := range endpoints {
+		endpoint = endpointWithPort(endpoint)
+
+		if _, exists := e.exporters[endpoint]; !exists {
+			cfg := e.exporterFactory.CreateDefaultConfig()
+			oCfg := cfg.(*otlpexporter.Config)
+
+			// is this wrong? are we overriding default fields, or is the config unmarshaller taking care of it?
+			oCfg.ExporterSettings = e.config.Protocol.OTLP.ExporterSettings
+			oCfg.TimeoutSettings = e.config.Protocol.OTLP.TimeoutSettings
+			oCfg.QueueSettings = e.config.Protocol.OTLP.QueueSettings
+			oCfg.RetrySettings = e.config.Protocol.OTLP.RetrySettings
+			oCfg.GRPCClientSettings = e.config.Protocol.OTLP.GRPCClientSettings
+
+			oCfg.Endpoint = endpoint
+			exp, err := e.exporterFactory.CreateTracesExporter(ctx, e.templateCreateParams, cfg)
+			if err != nil {
+				e.logger.Warn("failed to create new trace exporter for endpoint", zap.String("endpoint", endpoint))
+				continue
+			}
+			e.exporters[endpoint] = exp
+		}
+	}
+}
+
+func (e *exporterImp) removeExtraExporters(ctx context.Context, endpoints []string) {
+	for existing := range e.exporters {
+		if !endpointFound(existing, endpoints) {
+			e.exporters[existing].Shutdown(ctx)
+			delete(e.exporters, existing)
+		}
+	}
+}
+
+func endpointFound(endpoint string, endpoints []string) bool {
+	for _, candidate := range endpoints {
+		if candidate == endpoint {
+			return true
+		}
+	}
+
+	return false
+}
+
 func (e *exporterImp) Shutdown(context.Context) error {
+	e.stopped = true
+	e.shutdownWg.Wait()
 	return nil
 }
 
 func (e *exporterImp) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
-	return nil
+	e.updateLock.RLock()
+	defer e.updateLock.RUnlock()
+
+	traceID := traceIDFromTraces(td)
+	endpoint := e.ring.endpointFor(traceID)
+	exp, found := e.exporters[endpoint]
+	if !found {
```

Comment:
Instead of `endpointFor`, can it return `exporterFor`, possibly with a restructure of the code, to avoid this case? I guess it's truly impossible, meaning we're missing an association in our code to get that safety at compile-time instead of runtime.

Reply:
e.g., instead of an EndpointRing, it could possibly be an ExporterRing

Reply:
This is because of the failure of building an exporter, as discussed in another comment, right? Given my last comment there, I wouldn't change the ring to be an exporter one instead of string, as that condition is quite extreme and unlikely to happen. If we do decide that the ring should only have healthy entries, we could add a callback to it, so that it's notified when entries have to be removed.